### DISCUSSION TITLE
Welcome to Auteur Discussions!

### DISCUSSION THREAD

**Discussion by erikstmartin** [microsoft/auteur#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by SonnyWalkman** [comment]
This project is awesome! It's close to what I need for a project I'm working on. I'm new to rust and have spent weeks getting to understand how rust runs under the hood. I've been working with python with gstreamer Glib for months and are interested in rust implementation.

Few questions if I may?

1. Can RTP import source be added?

2. How do you switch between source nodes? 

3. On source node, is Auteur able to unlink source, change location, then reattach the same disconnected pad? 
I like to swap between multiple static png's using imagefreeze. Basically disconnect source pad, sources fallbacks to base plate, change source location/uri e.g colourbars.png then re attach pad to reinstate node showing new png?

4. Can a tee be added to the pipeline to allow multiple destinations?

RTMP in parallel with mpegtsmux TS pipe to dvbmuxer as an example.

5. 
Can cue start time and duration be disabled. Is there a switch as I don't need to cue sources. Straight cut to sources.

**Reply by MathieuDuponchelle** [reply]
Hey @SonnyWalkman , thanks for posting here, we're curious what people (plan to) do with this tool :)

> Can RTP import source be added?

For now source nodes don't have the notion of "source family" as destination nodes do, that is because using URIs is flexible enough for the use cases we developed the project for (eg. you can pass `web:///<url>` and use a website as a regular source if you have a recent enough `cefsrc` on your path).

Ideally, you could do away with a specific source node family and instead use `rtpsrc` for your purpose, it implements `GstUriHandler` as well but I've never used that feature :)

> How do you switch between source nodes?

A robust approach is to use an intermediary mixer node, connect source nodes to it and schedule them for playback one after the other. In the future `auteur` might expose an "atomic" reconnection feature, but that isn't on the roadmap for now.

> On source node, is Auteur able to unlink source, change location, then reattach the same disconnected pad?

No, the URI is a construct-only feature of source nodes.

> I like to swap between multiple static png's using imagefreeze. Basically disconnect source pad, sources fallbacks to base plate, change source location/uri e.g colourbars.png then re attach pad to reinstate node showing new png?

You should be able to implement this using an intermediary mixer node. Note that mixers support base plates of their own :)

> Can a tee be added to the pipeline to allow multiple destinations?

You can think of producer nodes (sources, mixers) as tees, they can be connected to any number of consumer nodes (mixers, destinations) you want.

> Can cue start time and duration be disabled. Is there a switch as I don't need to cue sources. Straight cut to sources.

This is already supported: https://microsoft.github.io/auteur/auteur_controlling/controller/enum.GraphCommand.html#variant.Start

**Reply by SonnyWalkman** [reply]
Matthieu,
Thank-you for your detailed answers.
I'm really excited with this project.  It covers 80% of what I require. I'm new to rust programming and I find the code is very well modulised, and easy to adapt and add features.

Could a overlay feature be added to the mixer video pipeline to allow overlaying dynamic context using wpesrc for example?

Sonny.

Great work!!!

**Reply by MathieuDuponchelle** [reply]
Hey, thanks for the kind words, it's nice to get validation for the design @erikstmartin and I settled on. For what it's worth I'm also (relatively) new to rust programming, I do have a great mentor though :)

Regarding the overlay feature, the current design already allows for that, though I did need a few patches here and there to make the following example work as expected:

* https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/merge_requests/1239
* https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/merge_requests/538
* https://github.com/microsoft/auteur/pull/4

Here's one way to go about this:

```
import argparse

from node_schedule import create_mixer, create_rtmp_destination, start_node, connect, schedule_source, add_control_point, later, get_info

if __name__ == '__main__':
    parser = argparse.ArgumentParser('web-overlay')

    parser.add_argument('source_uri')

    args = parser.parse_args()

    source_uri = args.source_uri

    create_mixer('channel-1', 1920, 1080, 44100)
    create_rtmp_destination('centricular-output', 'rtmp://some-endpoint')
    start_node('centricular-output')
    connect('channel-1', 'centricular-output')
    start_node('channel-1')

    source_slot = schedule_source(source_uri, 'source_slot', 'channel-1', later(0))
    cef_slot = schedule_source('web://webkit.org/blog-files/3d-transforms/poster-circle.html', 'cef_slot', 'channel-1', later(0))

    add_control_point('control-1', source_slot, 'video::zorder', later(0), 1, interpolate=False)
    add_control_point('control-1', cef_slot, 'video::zorder', later(0), 2, interpolate=False)
```

The output looks like this:


https://user-images.githubusercontent.com/1278273/127778558-18753637-817e-494b-9029-7a988679e043.mp4

**Reply by SonnyWalkman** [reply]
Hey Matthieu,
What you have done is really impressive in such a so short time! Well done hats off to you.

I've done a few bits a pieces using python as a prototype however the performance on one computer is taxing.  
Rust blows the doors off Python in performance from the little I've done. 

I have a need for my project to receive Transport stream using the dvbsrc element selecting a program using PID.  Basically same as DVBSnoop .
The DVB receive and pushing the TS  is done on a separate X86 embedded PC. I require to get the selected TS across the network (LAN) into auteur. I was proposing using a simple  udpsink or tcpsink since the stream is already in a ts container and develop a source via these actrix actors to auteur.  The source to demux and decode the stream. Example:
1.  tsdemux=>decodebin3=>appsink  for both V and A. I also need to push headend stats via TCP socket to my app to produce a On-Screen-Display hence with overlay.

In addition, a destination sink to generate a transport stream. i.e ( encode video AYUV => H.264 => mpegtsmux ) & (encode audio RAW => aac => mpegtsmux ) => add SI Tables into mstp =>udpsink.

What do you think?

I haven't found much with rust using mpegtsmux element.

Do you have any skeleton code for a each source and destination  as a template to add pipeline elements and set properties etc to attach to auteur?

**Reply by MathieuDuponchelle** [reply]
In your case, I do not think you need a new source type, as you can specify a URI and let fallbacksrc / uridecodebin3 take care of the rest eg:

``` shell
gst-launch-1.0 filesrc location=<some-transport-stream.ts> ! tsparse set-timestamps=true alignment=7 ! queue ! udpsink host=127.0.0.1 port=5000
```

``` shell
gst-launch-1.0 playbin uri=udp://@127.0.0.1:5000
```

Regarding the destination, you could use https://github.com/microsoft/auteur/blob/main/server/src/destination.rs#L272-L351 as a template, note that I haven't yet designed an interface for fine-grained control of the encoders / muxers, those properties are currently hardcoded for each destination family, and I do not want to expose redundant controls per destination.

In the future, we will probably want to either use serialized [`GstEncodingProfiles`](https://gstreamer.freedesktop.org/documentation/pbutils/encoding-profile.html?gi-language=c) or roll up our own version of those, and either expect a single extra `profile` field in the configuration for relevant destination families, or possibly implement a new "Encoder" node type, this will require careful design however :)

I suggest we continue this discussion in a new Q&A discussion, also note that we (centricular) can offer consulting services if you have budget and want us to extend the project to fit your needs :)

**Reply by philn** [reply]
> Could a overlay feature be added to the mixer video pipeline to allow overlaying dynamic context using wpesrc for example?
> 

Chiming in to answer this question, as it was not answered by Mathieu. Yes `wpesrc` can be used as well, in case your setup doesn't have enough hw resources for CEF ;) You would need this MR though, in order to have `wpesrc` auto-plugged by the `uridecodebin3` element used by Auteur. https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/2498

In case you have any doubts about WPE/GstWPE don't hesitate to reach us out over at igalia.com ;)

Philippe

**Reply by MathieuDuponchelle** [reply]
Hey @philn I meant to check the code but since you're here :) What protocol did you pick for the URI? In cefsrc I went for both `web://` and `cef://`, I reckon we would want `web://` as well in wpesrc, what do you think?

**Reply by MathieuDuponchelle** [reply]
Also, curious if you're using `auteur` at all, if so it is my understanding that you have to introduce yourself in this thread :smile:

**Reply by philn** [reply]
> Hey @philn I meant to check the code but since you're here :) What protocol did you pick for the URI? In cefsrc I went for both `web://` and `cef://`, I reckon we would want `web://` as well in wpesrc, what do you think?

I use `wpe://` but yeah I suppose we could add `web://` as well.

Anyway, I was interested to test this project and have a look around :) I locally added RTSP destination support but given previous comments here I'm not sure it's worth upstreaming?

**Reply by MathieuDuponchelle** [reply]
> I use wpe:// but yeah I suppose we could add web:// as well.

I think it's nice, then if both wpesrc and cefsrc are present on the system the user can manipulate ranks

> Anyway, I was interested to test this project and have a look around :) I locally added RTSP destination support but given previous comments here I'm not sure it's worth upstreaming?

Using `rtspclientsink`? It does make sense to upstream new destination families, the part I'd rather keep generic / URI-based is sources :)

**Reply by philn** [reply]
Yes, with `rtspclientsink`.

**Reply by MathieuDuponchelle** [reply]
I see, then yeah feel free to propose, fwiw we'll need to expose some sort of encoding profile-like configuration option, but for now I don't want to commit to anything :)

**Reply by SonnyWalkman** [reply]
Well done.. Very impressive.

**Comment by SonnyWalkman** [comment]
Is there a way of using the source with a static file and switch between images?

 gst-launch-1.0 filesrc location=some.png ! decodebin ! imagefreeze ! autovideosink

Using a URI it seems its is relatively easy to swap between uri's. Can this be done using fielsrc  and videotestsrc and audiotestsrc etc? 

In addition, is the fallbackswitch available as a general element in Gstreamer or is it proprietary to Rust?

**Comment by SonnyWalkman** [comment]
Is there any plans to add features to Auteur? Could DeckLink destination be added to allow for monitoring in headless mode using SDI??
No Display..

In addition, could Auteur run on ARM Nvidia hardware Jetson or Xavier to utilise the GPU hardware encoders?

**Reply by MathieuDuponchelle** [reply]
Hey, no firm plans, at the moment I only have a branch I'm toying with to support connecting audio / video independently. I'm completely OK with adding a decklink estination family (probably using the new `decklinksink`). As for hardware encoders, I'd be open to implementing better encoder selection, I'm not quite sure whether I'd want API to control it at this stage, and what that API would look like if I did :)

**Comment by SonnyWalkman** [comment]
Mathieu,
Great to hear from you..
I understand your possibly busy with other projects at the moment.
Thank you for taking the time and considering decklink as a destination
output on auteur.
I have a Duo pcie card and was wanting to use the card SDI outputs to
preview 2 outputs to monitor 2 seperate rtmp feeds.

I’m putting together a project which is a basically a tv repeater which
monitors two dvb-s channels via a PCIe Dektec or TBS satellite card. When
there is not valid TS received, the input shows either a static image of
dynamic mp4 which loops.
 I was wanting to extend the source.rs using dvbsrc to do the dvb side of
things with a pipeline to  demuxing,  spilt  video and audio on desired
pids snd decoding using decodebin2 or decodebin3, deinterlace before
heading into auteur as 1280x720 raw video and audio to be switched. Here
your wpe addon use to do lower thirds and then through destination.rs
 encode using hardware GPU to be coupled with declink, rtmp into a TS
stream ready to be is multiplexed with another TS to produce a dvb
compliant TS to be transmitted using SDR or hides UM100-T usb dvbt
transmitter.
The dvb stream requires SI to be added. I was hoping mpegtsmux is able to
do all the hard work.


Rob
